File: credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long card_number;
    long x = 0;
    long last_digit;
    long product_of_multiplied = 0;
    long product_of_not_multiplied = 0;
    long card_identifier;
    long checksum;
    bool flag =
        false; // used for setting a flag if it is the digit that needs to be multiplied or not
    void validate(long checksum, long card_identifier);

    do
    {
        card_number = get_long("Card number: ");
    }
    while (card_number < 1);

    card_identifier = card_number;

    while (card_number > x)
    {

        last_digit = card_number % 10;

        if (flag == false)
        {

            product_of_not_multiplied += last_digit;
            flag = true;
            card_number = card_number / 10;
            continue;
        }

        else if (flag == true)
        {

            last_digit = last_digit * 2;

            // handling 2 digit number after multiplication
            // it could be at most 9*2=18

            if (last_digit >= 10)
            {
                long z = last_digit % 10;
                long y = last_digit / 10;

                last_digit = z + y;
            }
            product_of_multiplied += last_digit;
            flag = false;
            card_number = card_number / 10;
            continue;
        }
    }

    checksum = product_of_not_multiplied + product_of_multiplied;

    validate(checksum, card_identifier);
}

void validate(long checksum, long card_identifier)
{
    if (checksum % 10 == 0 &&
        (card_identifier / 1000000000000000 == 4 || card_identifier / 1000000000000 == 4))
    {
        printf("VISA\n");
    }

    else if (checksum % 10 == 0 &&
             (card_identifier / 10000000000000 == 34 || card_identifier / 10000000000000 == 37))
    {
        printf("AMEX\n");
    }
    else if (checksum % 10 == 0 &&
             (card_identifier / 100000000000000 == 51 || card_identifier / 100000000000000 == 52 ||
              card_identifier / 100000000000000 == 53 || card_identifier / 100000000000000 == 54 ||
              card_identifier / 100000000000000 == 55))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: credit.py
from cs50 import get_int, get_string
import math


def main():
    while True:
        card_number = get_string("Card number: ")
        if int(card_number) > 0:
            break

    checksum = luhns_algorithm(card_number)
    validate(checksum, card_number)


def luhns_algorithm(card_number):
    sum_of_multiplied = 0
    sum_of_not_multiplied = 0

    for element in str(card_number[::-2]):
        sum_of_not_multiplied = sum_of_not_multiplied + int(element)

    card_number = card_number[:-1]

    for element in str(card_number[::-2]):
        digit = int(element) * 2
        if digit > 9:
            digit = (digit % 10) + 1

        sum_of_multiplied = sum_of_multiplied + digit

    checksum = sum_of_not_multiplied + sum_of_multiplied

    if checksum % 10 == 0:
        return True
    else:
        return False


def validate(checksum, card_number):

    first_digit = card_number[0]
    second_digit = card_number[1]

    if checksum == False:
        print("INVALID")

    while checksum:
        if first_digit == "4" and (len(card_number) == 13 or len(card_number) == 16):
            print("VISA")
            break

        elif first_digit == "3" and (second_digit == "4" or second_digit == "7"):
            print("AMEX")
            break

        elif first_digit == "5" and (second_digit > "0" and second_digit < "6"):
            print("MASTERCARD")
            break

        else:
            print("INVALID")
            break


main()